### PR TITLE
NXP backend: MLPerf Tiny Anomaly detection enable test model

### PR DESCRIPTION
**Pull request by roman-janik-nxp**
### Summary
Enable and refactor MLPerf Tiny Anomaly detection tests.

### Test plan
Unit test provided.

cc @robert-kalmar @JakeStevens @digantdesai @rascani

### CODE REVIEW

**Comment by roman-janik-nxp** [comment]
Internal test: https://bamboo3.sw.nxp.com/browse/MLTECE-EXGHSPEC48-2

File: backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    dataset_creator = FromCalibrationDataDatasetCreator(
+        dataset, num_examples=num_samples, idx_to_label=labels
+    )
+
+    input_spec = ModelInputSpec(anomaly_detection.input_shape)
+    quant_type_key = "QAT" if use_qat else "PTQ"
+
+    mse = BOUNDS_MSE[quant_type_key]
+    comparator = NumericalStatsOutputComparator(max_mse_error=mse)
```
Comment (novak-vaclav) [inline backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py:69]: nit: Anomaly detection is technically classification, so there should be `classification_task=True` and perhaps `use_softmax=True`, however that depends on the architecture of the model.

File: backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    dataset_creator = FromCalibrationDataDatasetCreator(
+        dataset, num_examples=num_samples, idx_to_label=labels
+    )
+
+    def _get_class_from_reconstruction_error(
+        preds: np.ndarray,
+        pred_path: str,
+    ) -> np.ndarray:
+        sample_name = pred_path.split("/")[-2]
```
Comment (novak-vaclav) [inline backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py]: I'd move `_get_class_from_reconstruction_error` to `MLPerfTinyAnomalyDetection` class and add a comment why the reconstruction and the whole dataset flattening trick/workaround is necessary. I understand why it's necessary, but I know it only because we talked about it in the office.

It may not be possible because of the import error in OSS CI. Solution for this import problem might be to move `_get_class_from_reconstruction_error` to  `MLPerfTinyAnomalyDetection` with argument `input_path` and import the `OUTPUTS_DIR` here in this test and use it to construct `input_path`.
If you decide to leave `_get_class_from_reconstruction_error` where it is, also add a comment about the CI limitation.

File: backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        test_name = get_test_name(request)
+        input_path = os.path.join(
+            OUTPUTS_DIR,
+            test_name,
+            "dataset/calibration/",
+            sample_name,
+        )
+        inps = np.fromfile(input_path, dtype=preds.dtype).reshape(preds.shape)
+        threshold = 17.0  # Empirically chosen
```
Comment (novak-vaclav) [inline backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py]: nit: If you choose to move the helper to `MLPerfTinyAnomalyDetection` class, please extract the `threshold` into class static variable.

File: examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                )
+
+            flat_xs.extend(list(input_batch))
+            flat_ys.extend([single_label] * batch_size)
+
+        self._dataset_flattened = True
+        self.dataset.examples = list(zip(flat_xs, flat_ys))
+
+    def _init_dataset(self) -> Dataset:
+        if self._use_random_dataset:
```
Comment (novak-vaclav) [inline examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py:111]: Why not perform dataset flattening once and for all in the `_init_dataset`? Then you could remove the `_dataset_flattened` class flag.

File: examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+from executorch.examples.models.mlperf_tiny import DeepAutoEncoderModel
+from executorch.examples.nxp.models.mlperf_tiny.mlperf_tiny_model import MLPerfTinyModel
+from torch.utils.data import Dataset
+from torchao.quantization.pt2e import disable_observer
+from tqdm import tqdm
+
+log = logging.getLogger(__name__)
+
+
+class MLPerfTinyAnomalyDetection(MLPerfTinyModel):
```
Comment (novak-vaclav) [inline examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py:28]: I believe the whole flattening logic introduced a case where the previously discussed class hierarchy (`MLPerfTinyModel` -> `MLPerfTinyAnomalyDetection (executorch)` -> `MLPerfTinyAnomalyDetection (executorch-integration)`) makes more sense than the current state, since the code and logic redudancy would become a much larger issue.
If you decide to move the `_get_class_from_reconstruction_error` helper here, imo you have two ways to approach this refactor:
a) rework the hierarchy to `MLPerfTinyModel` -> `MLPerfTinyAnomalyDetection (executorch)` -> `MLPerfTinyAnomalyDetection (executorch-integration)` (and in image classification, kws and visual wake words as well to make it uniform) and then include the flattening logic only in `MLPerfTinyAnomalyDetection (executorch)`
b) extract the flattening to some global helper function

**Review by novak-vaclav — COMMENTED** [review]
I've got a few comments, mainly regarding documentation and the implementation of dataset flattening.
Otherwise LGTM 😊👌

File: backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py (hunk 1 above)
Comment (roman-janik-nxp) [reply backends/nxp/tests/models/test_mlperf_tiny_anomaly_detection.py:69]: No, just the output is interpreted as a class. Adding `is_classification_task=True` gives some number in the table, but it is not a class as the model doesn't produce a class output - it doesn't have Softmax. The output is an anomaly score, that is then interpreted as class according to a threshold.

File: examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py (hunk 4 above)
Comment (roman-janik-nxp) [reply examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py:111]: The internal dataset generated with batch_size=98. The `DatasetCreator` reads it and creates example with shape `(98, 640)`. If it is flattened right away, there will be shape mismatch at the model input, as the model accepts `(98, 640)` input. The internal trained model is already trained with that shape, changing would require new model and dataset. However, this leads to problems with QAT that was not available when the model was trained, that's why dataset flattening was needed.

File: examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py (hunk 5 above)
Comment (roman-janik-nxp) [reply examples/nxp/models/mlperf_tiny/anomaly_detection/mlperf_tiny_anomaly_detection.py:28]: Yes. here it makes more sense and the code would be duplicated a lot. I moved the `_get_class_from_reconstruction_error ` in the class, so I will introduce inheritance.

**Review by robert-kalmar — APPROVED** [review]
